client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssFetcher.java

```diff
@@ -0,0 +1,237 @@
+      RssFetcher.ShuffleErrors.WRONG_MAP.toString());
+    connectionErrs = reporter.getCounter(SHUFFLE_ERR_GRP_NAME,
+      RssFetcher.ShuffleErrors.CONNECTION.toString());
+    wrongReduceErrs = reporter.getCounter(SHUFFLE_ERR_GRP_NAME,
+      RssFetcher.ShuffleErrors.WRONG_REDUCE.toString());
+
+    this.shuffleReadClient = shuffleReadClient;
+    this.totalBlockCount = totalBlockCount;
+  }
+
+  public void fetchAllBlocks() throws IOException, InterruptedException {
+    while (!stopped) {
+      try {
+        // If merge is on, block
+        merger.waitForResource();
+        // Do shuffle
+        metrics.threadBusy();
+        copyFromRssServer();
+      } catch (Exception e) {
+        stopFetchAfterException(e);
+      } finally {
+        metrics.threadFree();
+      }
+    }
+  }
+
+  @VisibleForTesting
+  public void copyFromRssServer() throws IOException {
+    // fetch a block
+    final long startFetch = System.currentTimeMillis();
+    CompressedShuffleBlock compressedBlock = shuffleReadClient.readShuffleBlockData();
+    ByteBuffer compressedData = null;
+    if (compressedBlock != null) {
+      compressedData = compressedBlock.getByteBuffer();
+    }
+    long fetchDuration = System.currentTimeMillis() - startFetch;
+    readTime += fetchDuration;
+
+    // uncompress the block
+    if (compressedData != null) {
+      final long startDecompress = System.currentTimeMillis();
+      ByteBuffer uncompressedData = RssShuffleUtils.decompressData(
+        compressedData, compressedBlock.getUncompressLength());
+      unCompressionLength += compressedBlock.getUncompressLength();
+      long decompressDuration = System.currentTimeMillis() - startDecompress;
+      decompressTime += decompressDuration;
+
+      // Allocate a MapOutput (- either in-memory or on-disk) to put uncompressed block
+      final long startSerialization = System.currentTimeMillis();
+      TaskAttemptID mapId = MRRssUtils.createMRTaskAttemptId(
+        reduceId.getJobID(), TaskType.MAP, compressedBlock.getBlockId());
+      MapOutput<K, V> mapOutput = null;
+      try {
+        mapOutput = merger.reserve(mapId, compressedBlock.getUncompressLength(), 0);
+      } catch (IOException ioe) {
+        // kill this reduce attempt
+        ioErrs.increment(1);
+        throw ioe;
+      }
+      // Check if we can shuffle *now* ...
+      if (mapOutput == null) {
+        LOG.info("RssMRFetcher" + " - MergeManager returned status WAIT ...");
+        //Not an error but wait to process data.
+        return;
+      }
+
+      // Write and commit uncompressed data to MapOutput.
+      // In the majority of cases, merger allocates memory to accept data,
+      // but when data size exceeds the threshold, merger can also allocate disk.
+      // So, we should consider the two situations, respectively.
+      try {
+        if (mapOutput instanceof InMemoryMapOutput) {
+          InMemoryMapOutput inMemoryMapOutput = (InMemoryMapOutput) mapOutput;
```

Comment:
put the judge logic into RssBypassWriter

Reply:
OK